### PR TITLE
sg: add pidfile

### PR DESCRIPTION
**Pull request by mrnugget**
TODOs:

- [x] Do not stop process startup if processes in pid files are dead (because they crashed and didn't clean up pidfile)

## Test plan

- Manually tested during SG hacking hour 
- (@jhchabran) Manually tested the pid purging scenario
  - `sg start otel` 
  - `cp ~/.sourcegraph/sg.pid.*.json ~`
  - kill sg 
  - `ls ~/.sourcegraph/sg.pid.*.json` got nothing, good 
  - `cp ~/sg.pid.*.json ~/.sourcegraph/` 
  - `sg start otel` saw my `println` saying it got trashed 
  - starts fine even with the existing file (and a new one is created).

### CODE REVIEW

**Comment by jhchabran** [comment]
cc @mrnugget

File: dev/sg/internal/run/run.go [hunk 1]
```diff
@@ -28,6 +31,112 @@ import (
...
+		if slices.Equal(content.Args, args) {
+			// If a pid already exists, let's check if it's still running.
+			alive, err := isPidAlive(int32(content.Pid))
+			if err != nil {
+				return 0, true, errors.Wrapf(err, "could not check if pid %d is alive", content.Pid)
+			}
+			if !alive {
+				// Trash the pidfile and return that it's not running.
+				_ = os.Remove(match)
+				return 0, false, nil
```
Comment (mrnugget) [inline dev/sg/internal/run/run.go:72]: You can't return here - what if the next `match` _is_ alive?

File: dev/sg/internal/run/run.go [hunk 2]
```diff
@@ -28,6 +31,112 @@ import (
...
+
+	if err := os.WriteFile(pidFilePath, b, 0644); err != nil {
+		return err
+	}
+
+	interrupt.Register(func() { os.Remove(pidFilePath) })
+
+	return nil
+}
+
```
Comment (mrnugget) [inline dev/sg/internal/run/run.go:139]: Now that we have three functions: how about moving all of this to `run/pid.go`?

File: dev/sg/internal/run/run.go [hunk 3]
```diff
@@ -128,6 +241,8 @@ type cmdRunner struct {
 
 	repositoryRoot string
 	parentEnv      map[string]string
+
+	processes map[string]int
```
Comment (mrnugget) [inline dev/sg/internal/run/run.go:245]: This can be removed

**Review by mrnugget — COMMENTED** [review]
Left some comments. I can also pick these up.

**Comment by marekweb** [comment]
Nice. Is this mainly to avoid the annoyance from accidentally running multiple `sg start`s or was there a bigger issue that inspired it?

Could process ID reuse be a problem? For example: we crashed and another process is later running with the same ID? I'm guessing that it's not an issue in practice but I'm curious if that's a thing that can happen.

**Comment by mrnugget** [comment]
> For example: we crashed and another process is later running with the same ID? I'm guessing that it's not an issue in practice but I'm curious if that's a thing that can happen.

It would check whether the process is still running and remove the file if not.

**Pull request merged**